### PR TITLE
Add multi disambiguation for positional scalars

### PR DESCRIPTION
**Pull request by ab5tract**
Until now there has been no way to capture the use of `$[]`
(as opposed to `[]`) by the caller of a multi routine.

For example:
```
    multi sub($a where *.can("elems") == 2) {...}
    multi sub(@a) {...}

    a $[1,2]
```

In this case, the `@a` candidate always wins (the (ugly) where clause is never
tested).

Or consider:
```
    multi sub(Array $a) {...}
    multi sub(@a) {...}

    a $[1,2]; # goes to first
    a [1,2];  # also goes to first
```

Finally:
```
    multi sub(Positional $a) {...}
    multi sub(@a) {...}

    a $[1,2]; # failure due to ambiguous dispatch
```

This patch addresses the issue by removing the ambiguity from this
third example. By specifying a scalar argument to be of type Postional,
it will "win" against an @ argument in the same place in a competing
multi signature.

By taking this approach, it is hoped that the fallout to the ecosystem
will be minimal as it is unlikely that anyone has intentionally shipped
multis that fail to dispatch due to ambiguity.

Note: This patch does not address signatures where the positional scalar
utilizes a where-guard. Since the where-guard already breaks the tie that
is at the heart of this patch, it is much more likely that such a change
would adversely affect the ecosystem.

### CODE REVIEW

File: src/core.c/Array.rakumod [hunk 1]
```diff
@@ -1178,6 +1229,7 @@ my class Array { # declared in BOOTSTRAP
     multi method splice(Array:D:
       Int:D $offset, Whatever $, @new
     --> Array:D) {
+        say "in the array candidate";
```
Comment (lizmat) [inline src/core.c/Array.rakumod:1232]: Debug leftover?

File: src/vm/moar/dispatchers.nqp [hunk 2]
```diff
@@ -2574,6 +2599,69 @@ sub raku-multi-plan(
...
+            while $w < $total-capture-scalar-pos-args {
+                $perfect-match := $perfect-match
+                    && nqp::iseq_i(nqp::atpos_i($capture-scalar-pos-args, $w), nqp::atpos_i($cand-scalar-pos-params, $w));
+                ++$w;
+            }
+            return [ %current ] if $perfect-match;
+        }
+        ++$i;
+    }
+    return [ %current ];
```
Comment (lizmat) [inline src/vm/moar/dispatchers.nqp:2662]: There's an ever so slight overhead in using `return`: don't use it in hot code unless it's needed

File: src/core.c/Array.rakumod [hunk 3]
```diff
@@ -1093,6 +1093,57 @@ my class Array { # declared in BOOTSTRAP
                :range("0..^{self.elems - $offset}")
              ).throw
     }
+    #------ splice(offset,size,positional scalar) candidates
```
Comment (lizmat) [inline src/core.c/Array.rakumod:1096]: would prefer the `Array.splice` changes to be in a separate PR, after it has become clear that the `Positional $foo` feature works and has spectests.  This would allow for better bisectability should problems with `Array.splice` becime clear after a blin run.

File: src/core.c/Array.rakumod (hunk 1 above)
Comment (ab5tract) [reply src/core.c/Array.rakumod:1232]: Good catch. I thought I nuked all those.

File: src/core.c/Array.rakumod (hunk 3 above)
Comment (ab5tract) [reply src/core.c/Array.rakumod:1096]: They are already separate commits, but if you feel a second PR is better I don't have a problem with that.

File: src/vm/moar/dispatchers.nqp (hunk 2 above)
Comment (ab5tract) [reply src/vm/moar/dispatchers.nqp]: Fixed.

File: src/core.c/Array.rakumod (hunk 3 above)
Comment (ab5tract) [reply src/core.c/Array.rakumod]: The PR is now adjusted to only address the positional scalar disambiguation.

**Comment by lizmat** [comment]
A more general comment: without identical changes to the non-dispatch case, would this create different semantics on the JVM backend?

**Comment by ab5tract** [comment]
> A more general comment: without identical changes to the non-dispatch case, would this create different semantics on the JVM backend?

Yes, unfortunately. Good pointing this out.

**Comment by ab5tract** [comment]
Legacy dispatch now supports the same proposed semantics.
